src/MongoDB.Driver.Core/Core/Servers/ServerMonitor.cs

```diff
@@ -102,6 +145,45 @@ public void RequestHeartbeat()
             }
         }
 
+        // private methods
+        private CommandWireProtocol<BsonDocument> InitializeIsMasterProtocol(IConnection connection)
+        {
+            BsonDocument isMasterCommand;
+            var commandResponseHandling = CommandResponseHandling.Return;
+            if (connection.Description.IsMasterResult.TopologyVersion != null)
+            {
+                // the final Readtimeout will be:
+                // connectTimeoutMS (the default heartbeat read timeout) + maxAwaitTimeMS (_heartbeatInterval)
+                connection.AddAdditionalReadTimeout(_heartbeatInterval);
```

Comment:
I left `AddAdditional` timeout instead `SetRead` (fully new) timeout because I had problems with finding the right way to do it. If the current way is not good enough, I will change it

Reply:
I would like to find a way to use `SetReadTimeout` instead of `AddAdditionalReadTimeout`.

The approach with `AddAdditionalReadTimeout` is ill-defined because there is no way to know what the current value is, so who is to say what the end result of calling `AddAdditionalReadTimeout` is?

Let's calculate the actual timeout we want and call a deterministic method like `SetReadTimeout`.

Reply:
I'm fine with an API that includes a getter as well as a setter for the `ReadTimeout`.

I still think a statement like the following is not good:

    connection.ReadTimeout += somevalue;

Because it is non-deterministic. What is the previous value of `ReadTimeout`?

But at least that can be corrected without changing the API.

I don't understand why you can't just do something like this (WITHOUT `+=`):

    var newReadTimeout = computed from settings;
    connection.ReadTimeout = newReadTimeout;

If you don't have all the available settings then just make the missing settings available.